Detect stale loom-tools installation and warn or fail gracefully

## Problem

During shepherd orchestration for #1981, milestone events `phase_completed` and `judge_retry` were reported but the system logged:

```
[ERROR] Unknown event 'phase_completed'
[ERROR] Unknown event 'judge_retry'
```

These events were properly defined in the source code but the pipx-installed `loom-milestone` CLI was outdated and didn't recognize them.

## Root Cause

The loom-tools package was installed via pipx from the source directory. When the source was updated with new events, the installed version wasn't automatically updated.

The fix was: `pipx reinstall loom-tools`

## Impact

- Milestone events failed silently (logged error but orchestration continued)
- Progress tracking was incomplete
- Confusing error messages that didn't point to the real problem

## Proposed Solutions

### Option 1: Version check on startup
Add a version check in shepherd startup that compares:
- Installed loom-tools version/commit
- Source code version/commit

Warn if they differ.

### Option 2: Prefer source over installed
Change `loom-tools.sh` to prefer the source Python module (with PYTHONPATH) over the system-installed CLI. This ensures the latest code is always used.

Current priority in `run_loom_tool()`:
1. System-installed CLI ← prefers outdated
2. venv CLI
3. Python module with PYTHONPATH ← always current

Proposed priority:
1. Python module with PYTHONPATH (if in loom repo)
2. venv CLI
3. System-installed CLI

### Option 3: Fail on unknown events
Currently `report_milestone()` logs an error but returns `False`. The caller may not check this return value.

Make the shepherd fail hard on unknown milestone events so the problem is immediately visible.

## Acceptance Criteria

- [ ] Stale installation is detected before it causes confusing errors
- [ ] Clear message tells user how to fix (e.g., "Run: pipx reinstall loom-tools")
- [ ] Consider auto-reinstall option for development setups

--- a/defaults/scripts/lib/loom-tools.sh
+++ b/defaults/scripts/lib/loom-tools.sh
@@ -64,10 +64,13 @@ find_loom_tools() {
 #   $@ - Arguments to pass to the command
 #
 # Priority:
-#   1. System-installed CLI (loom-<name>)
+#   1. Python module with PYTHONPATH (if loom-tools source exists)
 #   2. venv CLI in loom-tools directory
-#   3. Python module with PYTHONPATH set
+#   3. System-installed CLI (loom-<name>)
 #   4. Error with helpful message
+#
+# This order ensures source code is always authoritative during development.
+# When running in a target repo (not the Loom source), falls back to installed CLI.
 run_loom_tool() {
     local cli_name="$1"
     local module_name="$2"
@@ -75,25 +78,25 @@ run_loom_tool() {
 
     local full_cli="loom-${cli_name}"
 
-    # Try system-installed CLI first
-    if command -v "$full_cli" >/dev/null 2>&1; then
-        exec "$full_cli" "$@"
-    fi
-
-    # Try to find loom-tools directory
+    # Try to find loom-tools directory (source takes precedence)
     local script_dir
     script_dir="$(cd "$(dirname "${BASH_SOURCE[1]}")" && pwd)"
 
     if find_loom_tools "$script_dir"; then
+        # Try Python module with PYTHONPATH first (always current source)
+        if [[ -d "$LOOM_TOOLS_SRC/loom_tools" ]]; then
+            PYTHONPATH="${LOOM_TOOLS_SRC}:${PYTHONPATH:-}" exec python3 -m "loom_tools.${module_name}" "$@"
+        fi
+
         # Try venv CLI
         if [[ -x "$LOOM_TOOLS_DIR/.venv/bin/$full_cli" ]]; then
             exec "$LOOM_TOOLS_DIR/.venv/bin/$full_cli" "$@"
         fi
+    fi
 
-        # Try Python module with PYTHONPATH
-        if [[ -d "$LOOM_TOOLS_SRC/loom_tools" ]]; then
-            PYTHONPATH="${LOOM_TOOLS_SRC}:${PYTHONPATH:-}" exec python3 -m "loom_tools.${module_name}" "$@"
-        fi
+    # Fall back to system-installed CLI (for target repos without source)
+    if command -v "$full_cli" >/dev/null 2>&1; then
+        exec "$full_cli" "$@"
     fi
 
     # Not found - provide helpful error message

--- a/loom-tools/tests/test_installation_verification.py
+++ b/loom-tools/tests/test_installation_verification.py
@@ -476,3 +476,164 @@ class TestInstalledScriptsMatchDefaults:
         assert not mismatches, (
             f"Scripts differ from defaults (may need reinstall): {mismatches}"
         )
+
+
+class TestSourcePriorityOverInstalled:
+    """Verify loom-tools.sh library prefers source over installed CLI.
+
+    Related issue: #2085 - loom-tools source not being picked up after modifications
+
+    The run_loom_tool() function in loom-tools.sh should prefer source Python
+    modules (via PYTHONPATH) over system-installed CLI commands to ensure
+    source code changes are immediately reflected during development.
+    """
+
+    @pytest.fixture
+    def defaults_dir(self) -> pathlib.Path:
+        return _find_defaults_dir()
+
+    @pytest.fixture
+    def repo_root(self) -> pathlib.Path:
+        return _find_repo_root()
+
+    def test_loom_tools_library_exists(self, defaults_dir: pathlib.Path) -> None:
+        """loom-tools.sh library should exist."""
+        lib = defaults_dir / "scripts" / "lib" / "loom-tools.sh"
+        assert lib.exists(), "loom-tools.sh library not found"
+
+    def test_source_priority_in_documentation(
+        self, defaults_dir: pathlib.Path
+    ) -> None:
+        """loom-tools.sh documentation should indicate source-first priority."""
+        lib = defaults_dir / "scripts" / "lib" / "loom-tools.sh"
+        content = lib.read_text()
+
+        # Find run_loom_tool function documentation specifically
+        func_start = content.find("run_loom_tool()")
+        assert func_start != -1, "run_loom_tool() function not found"
+
+        # Extract just the comment block before the function
+        func_section = content[:func_start].split("\n")
+        # Find the last "# Priority:" before the function
+        priority_section_start = None
+        for i, line in enumerate(func_section):
+            if "# Priority:" in line:
+                priority_section_start = i
+
+        assert priority_section_start is not None, (
+            "No Priority: documentation before run_loom_tool()"
+        )
+
+        # Extract priority lines (numbered items after "# Priority:")
+        priority_lines = []
+        for line in func_section[priority_section_start + 1:]:
+            stripped = line.strip()
+            if stripped.startswith("#") and any(
+                stripped.startswith(f"#   {n}.") for n in range(1, 10)
+            ):
+                priority_lines.append(line)
+            elif not stripped.startswith("#"):
+                break
+
+        assert len(priority_lines) >= 3, (
+            f"Incomplete priority documentation: {priority_lines}"
+        )
+
+        # First priority item should mention Python module or PYTHONPATH
+        first_priority = priority_lines[0].lower()
+        assert "python" in first_priority or "pythonpath" in first_priority, (
+            f"First priority should be Python/PYTHONPATH-based, got: {priority_lines[0]}"
+        )
+
+        # System CLI should NOT be first
+        assert "system" not in first_priority and "installed" not in first_priority, (
+            "System-installed CLI should not be first priority"
+        )
+
+    def test_source_check_before_system_cli(
+        self, defaults_dir: pathlib.Path
+    ) -> None:
+        """run_loom_tool() should check for source before system CLI."""
+        lib = defaults_dir / "scripts" / "lib" / "loom-tools.sh"
+        content = lib.read_text()
+
+        # Find the function body
+        func_start = content.find("run_loom_tool()")
+        assert func_start != -1, "run_loom_tool() function not found"
+
+        # Extract function body (until next top-level function or EOF)
+        func_body_start = content.find("{", func_start)
+        func_body_end = content.find("\n}", func_body_start)
+        func_body = content[func_body_start:func_body_end]
+
+        # Check order: find_loom_tools check should come before "command -v"
+        pythonpath_check = func_body.find("PYTHONPATH")
+        find_tools_check = func_body.find("find_loom_tools")
+        command_check = func_body.find('command -v "$full_cli"')
+
+        assert find_tools_check != -1, "find_loom_tools() call not found"
+        assert pythonpath_check != -1, "PYTHONPATH exec not found"
+        assert command_check != -1, "command -v fallback not found"
+
+        # Source-based execution should come before system CLI check
+        assert find_tools_check < command_check, (
+            "find_loom_tools() should be checked before system CLI"
+        )
+        assert pythonpath_check < command_check, (
+            "PYTHONPATH exec should come before system CLI fallback"
+        )
+
+    def test_fallback_to_installed_cli_exists(
+        self, defaults_dir: pathlib.Path
+    ) -> None:
+        """run_loom_tool() should still fall back to installed CLI if no source."""
+        lib = defaults_dir / "scripts" / "lib" / "loom-tools.sh"
+        content = lib.read_text()
+
+        # The fallback should use "command -v" to check for installed CLI
+        assert 'command -v "$full_cli"' in content, (
+            "Fallback to system CLI not found"
+        )
+
+        # And it should exec the CLI if found
+        assert 'exec "$full_cli"' in content, (
+            "exec for system CLI not found"
+        )
+
+    def test_pythonpath_sets_correct_source(
+        self, defaults_dir: pathlib.Path
+    ) -> None:
+        """PYTHONPATH should be set to loom-tools source directory."""
+        lib = defaults_dir / "scripts" / "lib" / "loom-tools.sh"
+        content = lib.read_text()
+
+        # Should use LOOM_TOOLS_SRC in PYTHONPATH
+        assert "LOOM_TOOLS_SRC" in content, (
+            "LOOM_TOOLS_SRC variable not used"
+        )
+        assert 'PYTHONPATH="${LOOM_TOOLS_SRC}' in content, (
+            "PYTHONPATH not set from LOOM_TOOLS_SRC"
+        )
+
+    def test_venv_checked_after_source(
+        self, defaults_dir: pathlib.Path
+    ) -> None:
+        """venv CLI should be checked after direct source but before system CLI."""
+        lib = defaults_dir / "scripts" / "lib" / "loom-tools.sh"
+        content = lib.read_text()
+
+        func_start = content.find("run_loom_tool()")
+        func_body = content[func_start:]
+
+        pythonpath_exec = func_body.find("PYTHONPATH=")
+        venv_check = func_body.find('.venv/bin/"$full_cli"')
+        command_check = func_body.find('command -v "$full_cli"')
+
+        # Order should be: PYTHONPATH source > venv > system CLI
+        if venv_check != -1:  # venv check is optional
+            assert pythonpath_exec < venv_check, (
+                "PYTHONPATH source should be checked before venv"
+            )
+            assert venv_check < command_check, (
+                "venv should be checked before system CLI"
+            )

